File: javascript/53.MaximumSubarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const maxSubArray = nums => {
    let max = -1  * Math.abs(2 * nums[0])
    let current = -1 * Math.abs(2* nums[0])

    nums.forEach(val => {
        current += val
        if (val > current) current = val
        max = Math.max(max, current)
    })

    return max
}

/**
 *  @Runtime: {104 ms}
 *  @Memory: {49.2 MB}
 */

File: java/53.Maximum_Subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int max = -1* Math.abs(2 * nums[0]);
        int curr = max;

        for (int num : nums) {
            curr += num;
            if (num > curr) curr = num;
            max = Math.max(max, curr);
        }

        return max;
    }
}